File: old/leetcode/word-pattern.cpp
#include <vector>
#include <sstream>
#include <iostream>
#include <map>
using namespace std;

class Solution
{
private:
    vector<pair<string,int>> extractPattern(vector<string> sequence)
    {
        string prev = "";
        vector<pair<string,int>> pattern;
        for (const auto &c : sequence)
        {
            if (c != prev)
            {
                pattern.push_back({c, 1});
                prev = c;
            }
            else
            {   
                pattern[pattern.size() - 1].second++;
            }
        }
        return pattern;
    }

public:
    Solution() {

    }
    bool wordPattern(string pattern, string s)
    {
        vector<string> words;
        stringstream ss = stringstream(s);
        string word;
        while(ss >> word) {
            words.push_back(word);
        }
        
        map<char, string> lr;
        map<string, char> rl;
        vector<pair<string, int>> p = extractPattern(words);
        int i = 0;
        for (const auto& count: p) {
            if (i+count.second > pattern.size()) {
                return false;
            }

            char c = pattern[i];
            for (int j=1; j < count.second; j++) {
                if (pattern[i+j] != c) {
                    return false;
                }
            }
            
            if (lr.find(c) == lr.end()){
                lr[c] = count.first;
            } else if (lr[c] != count.first) {
                return false;
            }

            if (rl.find(count.first) == rl.end()) {
                rl[count.first] = c;
            } else if (rl[count.first] != c){
                return false;
            }

            i += count.second;
        }

        if (i != pattern.size()) {
            return false;
        }
        return true;
    }
};

int main() {
    Solution s = Solution();
    cout << s.wordPattern("abc", "dog cat dog");
}

File: old/leetcode/word-pattern.go
package main

import (
	"fmt"
	"strings"
)

type pair struct {
	word  string
	count int
}

func extractPattern(sentence string) []pair {
	pattern := make([]pair, 0)
	prev := ""
	for _, word := range strings.Split(sentence, " ") {
		if word != prev {
			pattern = append(pattern, pair{word, 1})
			prev = word
		} else {
			pattern[len(pattern)-1].count++
		}
	}
	return pattern
}

func wordPattern(pattern string, s string) bool {
	sentPattern := extractPattern(s)
	lr := make(map[byte]string)
	rl := make(map[string]byte)

	pi := 0
	for _, wordDetail := range sentPattern {
		if pi+wordDetail.count > len(pattern) {
			return false
		}

		char := pattern[pi]
		for j := pi + 1; j < pi+wordDetail.count; j++ {
			if pattern[j] != char {
				return false
			}
		}

		if word, ok := lr[char]; !ok {
			lr[char] = wordDetail.word
		} else if word != wordDetail.word {
			return false
		}

		if c, ok := rl[wordDetail.word]; !ok {
			rl[wordDetail.word] = char
		} else if c != char {
			return false
		}

		pi += wordDetail.count
	}

	if pi != len(pattern) {
		return false
	}
	return true
}

func main() {
	fmt.Println(wordPattern("abb", "dog cat cat dog"))
	// fmt.Println("abc", "dog cat dogs")
}
